### ISSUE TITLE
'<' not working

### ISSUE THREAD

**Issue by kenjitoyama** [PyUserInput/PyUserInput#15]
When I use tap_key('<') or press_key('<') I get a '>' instead. Everything else seems to be working.

I'm at commit 4229bdfe4a3ccbfc49a2600e826c79a5634297d0 running on Ubuntu:
> cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"

It happens for both python 2 (2.7.12) and 3 (3.5.1).

**Comment by pepijndevos** [comment]
That's a funny bug for a Sunday morning. It appears '<' has its own keycode, while '>' has the same keycode as '.'

    k.tap_key(k.lookup_character_keycode("<")) # <
    k.tap_key(k.lookup_character_keycode(">")) # .

But running this code does the expected thing, and presses ,/. or </> if you hold shift.

    k.tap_key(k.lookup_character_keycode(","))
    k.tap_key(k.lookup_character_keycode("."))

Down the rabbit hole we go...

**Comment by moses-palmer** [comment]
For some directions down in that hole, have a look at the [pynput implementation of key resolving](https://github.com/moses-palmer/pynput/blob/master/lib/pynput/keyboard/_xorg.py#L373) (using [this](https://github.com/moses-palmer/pynput/blob/master/lib/pynput/_util/xorg.py#L319) rather simple function).

It appears to work:

    moses@kafka:~$ python -c 'import pynput; pynput.keyboard.Controller().type("<>")'
    <>moses@kafka:~$ <>

**Comment by pepijndevos** [comment]
I didn't dig through your code too much, but you appear to do more or less the same as we do. That is, resolve "<" to "less" and use  `Xlib.XK.string_to_keysym()` to get the keysym.

I suspect your handling of the shift key is better though. Our code determines it needs to press shift for both < and >, while the keycode we get from `Xlib.display.Display.keysym_to_keycode()` does not need shift, and in fact pressing shift turns it into >.

Observe this curious code:
```
import Xlib
import Xlib.display
import Xlib.XK

d = Xlib.display.Display()

for s in ["greater", "less", "period", "comma"]:
    ks = Xlib.XK.string_to_keysym(s)
    kk = d.keysym_to_keycode(ks)
    kss = [d.keycode_to_keysym(kk, i) for i in range(4)]
    print s, ks, kk, kss
```

Which on my machine produces
```
greater 62 60 [46L, 62L, 46L, 62L]
less 60 94 [60L, 62L, 60L, 62L]
period 46 60 [46L, 62L, 46L, 62L]
comma 44 59 [44L, 60L, 44L, 60L]
```
As can be seen, less returns an oddly high keycode, that resolves back to the keysyms for greater and less, rather than less and comma, while greater shares the same keysyms as period.

**Comment by moses-palmer** [comment]
Ah, that seems to be very tied to some keyboard layout. I spent a lot of
time to abstract away layouts in pynput, and that was a huge task to
undertake on X.

**Comment by pepijndevos** [comment]
Right... the way modifier keys are implemented in this library is [dumb](https://github.com/PyUserInput/PyUserInput/blob/master/pykeyboard/base.py#L95). So how do you handle modifier keys in your project? If your library can type an euro sign with any degree of reliability, maybe it's time to start telling people to use pynput. (there is also #9 that is still unresolved)

**Comment by moses-palmer** [comment]
For X, the library maintains a means of mapping characters to keycodes and
modified that is initially retrieved by reading the layout.

Every time it is requested to handle a key with a graphic representation,
it searches for the corresponding keycode-modifier pair and sends an event.
If the character cannot be found, it locates an unused character in the
layout, adds the requested glyph at that position and sends the new pair.
That is the reason it can be used to type for example Russian or Thai even
though no characters of those alphabets are mapped in the layout.

X can barely be seen under all its warts.

**Comment by kenjitoyama** [comment]
Thanks, Pepijn and Moses for looking into this. I tried pynput and it does handle '<' correctly.

One difference between pynput and PyUserInput is that PyUserInput repeats the key if I don't issue a release on the key while pynput does not. I actually want this behavior in my application. How can I simulate it in pynput?

(We can take this to pynput's bug tracker if you prefer)

**Comment by moses-palmer** [comment]
In order to expose a higher level API, *pynput* must unfortunately use a lower level X API; *XFAKE* which is used by *PyUserInput* maintains some internal state that interferes with the ability to produce events for arbitrary glyphs. This has the unfortunate side effect that key repeating does not happen automatically.

This could be added to *pynput*; one idea is to use a single thread dispatching all *X* events, and using a `Queue` for communication---the read timeout could be used to repeat the last event if that was a press. This is however quite a bit of work, and since the *X* implementation already is very complex, I am somewhat reluctant to make it even more complex.

You can still manually trigger repeats:

    import pynput
    import time

    c = pynput.keyboard.Controller()

    for _ in range(5):
        c.press('c')
        time.sleep(0.5)
    c.release('c')
